src/target/riscv/riscv.c

```diff
@@ -462,9 +462,88 @@ static void trigger_from_breakpoint(struct trigger *trigger,
 	trigger->unique_id = breakpoint->unique_id;
 }
 
-static int maybe_add_trigger_t1(struct target *target,
-		struct trigger *trigger, uint64_t tdata1)
+static bool match_napot_checker(struct trigger *trigger, riscv_reg_t *tdata2)
+{
+	riscv_reg_t addr = trigger->address;
+	riscv_reg_t size = trigger->length;
+	if (size > 1 &&
+		(size & (size - 1)) == 0 &&
+		(addr & (size - 1)) == 0) {
+		if (tdata2)
+			*tdata2 = addr | ((size - 1) >> 1);
+		return true;
+	}
+	return false;
+}
+
+/* This variable is used to save the currently selected trigger,
+ * in order to reduce write TSELECT when call set_trigger
+ */
+static int selected_trigger;
+
+static int find_trigger(struct target *target, int type, bool chained, int *idx)
+{
+	RISCV_INFO(r);
+
+	int found;
+	int cnt = 0;
+	int num = chained ? 2 : 1;
+
+	for (unsigned i = 0; i < r->trigger_count; i++) {
+		if (r->trigger_unique_id[i] == -1) {
+			riscv_reg_t tdata1 = 0;
+			if (riscv_set_register(target, GDB_REGNO_TSELECT, i) != ERROR_OK)
+				return ERROR_FAIL;
+			selected_trigger = i;
+			if (riscv_get_register(target, &tdata1, GDB_REGNO_TDATA1) != ERROR_OK)
+				return ERROR_FAIL;
+			int t = get_field(tdata1, CSR_TDATA1_TYPE(riscv_xlen(target)));
+			if (type == t) {
+				found = (++cnt == num);
+				if (found) {
+					if (idx)
+						*idx = i - (num - 1);
+					return ERROR_OK;
+				}
+				continue;
+			}
+		}
+		cnt = 0;
```

Comment:
If you set cnt to 0 here, how will it ever become 2 on line 502?

Reply:
A continue on line 509 will skip the setting of 0 on line 511, and two consecutive t==type will make cnt equal to 2

Reply:
Got it. Thanks. Can you add a small comment to make it more obvious?